### PR TITLE
[api server] enable job spec server

### PR DESCRIPTION
**Pull request by Basasuya**
## Why are these changes needed?

add job spec server implement for job server

## Related issue number
https://github.com/ray-project/kuberay/issues/389

## Checks

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [*] Unit tests
   - [ ] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: apiserver/cmd/main.go [hunk 1]
```diff
@@ -58,6 +58,7 @@ func startRpcServer(resourceManager *manager.ResourceManager) {
 		grpc.MaxRecvMsgSize(math.MaxInt32))
 	api.RegisterClusterServiceServer(s, server.NewClusterServer(resourceManager, &server.ClusterServerOptions{CollectMetrics: *collectMetricsFlag}))
 	api.RegisterComputeTemplateServiceServer(s, server.NewComputeTemplateServer(resourceManager, &server.ComputeTemplateServerOptions{CollectMetrics: *collectMetricsFlag}))
+	api.RegisterRayJobServiceServer(s, server.NewRayJobServer(resourceManager))
```
Comment (Jeffwan) [inline apiserver/cmd/main.go:61]: Even collect metrics is not being used yet, let's keep the code structure same?

File: apiserver/pkg/client/job.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
+package client
+
+import (
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/ray-project/kuberay/apiserver/pkg/util"
+	rayclusterclient "github.com/ray-project/kuberay/ray-operator/pkg/client/clientset/versioned"
```
Comment (Jeffwan) [inline apiserver/pkg/client/job.go:8]: Let's rename it to `rayclient`?

File: apiserver/pkg/manager/client_manager.go [hunk 3]
```diff
@@ -10,6 +10,7 @@ import (
 
 type ClientManagerInterface interface {
 	ClusterClient() client.ClusterClientInterface
+	RayJobClient()  client.JobClientInterface
```
Comment (Jeffwan) [inline apiserver/pkg/manager/client_manager.go:13]: `RayJobClient` -> `JobClient ` for consistency

File: apiserver/pkg/manager/resource_manager.go [hunk 4]
```diff
@@ -30,6 +30,10 @@ type ResourceManagerInterface interface {
 	GetComputeTemplate(ctx context.Context, name string, namespace string) (*v1.ConfigMap, error)
 	ListComputeTemplates(ctx context.Context, namespace string) ([]*v1.ConfigMap, error)
 	DeleteComputeTemplate(ctx context.Context, name string, namespace string) error
+	CreateJob(ctx context.Context, apiCluster *api.Cluster) (*v1alpha1.RayJob, error)
+	GetJob(ctx context.Context, jobName string, namespace string) (*v1alpha1.RayJob, error)
+	ListJobs(ctx context.Context, namespace string) ([]*v1alpha1.RayJob, error)
```
Comment (Jeffwan) [inline apiserver/pkg/manager/resource_manager.go:35]: should we add `DeleteJob` as well?

File: apiserver/pkg/model/converter.go [hunk 5]
```diff
@@ -111,3 +116,24 @@ func FromKubeToAPIComputeTemplates(configMaps []*v1.ConfigMap) []*api.ComputeTem
...
+func FromCrdToApiJobs(jobs []*v1alpha1.RayJob) []*api.RayJob {
+	apiJobs := make([]*api.RayJob, 0)
+	for _, job := range jobs {
+		apiJobs = append(apiJobs, FromCrdToApiJob(job))
+	}
+	return apiJobs
+}
+
+func FromCrdToApiJob(job *v1alpha1.RayJob) *api.RayJob {
+	pbJob := &api.RayJob{
```
Comment (Jeffwan) [inline apiserver/pkg/model/converter.go:129]: I remember you proposed to add name and namespace? When you convert the object, should the information be added?

File: proto/job.proto [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+  // If jobId is not set, a new jobId will be auto-generated.
+  string job_id = 7;
+  // If set to true, the rayCluster will be deleted after the rayJob finishes
+  bool ShutdownAfterJobFinishes = 8;
+  // The label selector to choose exiting clusters
+  map<string, string> cluster_selector = 9;
+  // The cluster template
+  ClusterSpec cluster_spec = 10;
+  // logs
+  string logs = 11;
```
Comment (Jeffwan) [inline proto/job.proto:126]: we probably can remove logs now,  we either have a new method like `GET /apis/v1alpha2/namespaces/{namespace}/jobs/{name}/logs` or fetch it from somewhere else like  ELK or Ray Dashboard

File: proto/hack/generate.sh [hunk 7]
```diff
@@ -19,3 +19,13 @@ protoc -I. \
...
+
+# Generate a single swagger json file from the swagger json files of all models.
+# Note: use proto/swagger/{cluster,config,error}.swagger.json
+# Note: swagger files are generate to source folder directly. No files in ${TMP_OUTPUT}
+jq -s 'reduce .[] as $item ({}; . * $item) | .info.title = "KubeRay API" | .info.description = "This file contains REST API specification for KubeRay. The file is autogenerated from the swagger definition." | .info.version = "'0.3.0'" | .info.license = { "name": "Apache 2.0", "url": "https://raw.githubusercontent.com/ray-project/kuberay/master/LICENSE" }' \
+  /go/src/github.com/ray-project/kuberay/proto/swagger/cluster.swagger.json \
+  /go/src/github.com/ray-project/kuberay/proto/swagger/config.swagger.json \
+  /go/src/github.com/ray-project/kuberay/proto/swagger/error.swagger.json \
+  /go/src/github.com/ray-project/kuberay/proto/swagger/job.swagger.json \
+  > "/go/src/github.com/ray-project/kuberay/proto/kuberay_api.swagger.json"
```
Comment (Jeffwan) [inline proto/hack/generate.sh:31]: minor: add an empty line at the end

File: apiserver/pkg/util/job.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        apiJob.Name,
+			Namespace:   apiJob.Namespace,
+			Labels:      buildRayJobLabels(apiJob),
+			Annotations: buildRayJobAnnotations(apiJob),
+		},
+		Spec: rayalphaapi.RayJobSpec{
+			Entrypoint: apiJob.Entrypoint,
+			Metadata:  apiJob.Metadata,
+			RuntimeEnv: apiJob.RuntimeEnv,
```
Comment (Jeffwan) [inline apiserver/pkg/util/job.go:33]: rayalphaapi.RayJobSpec.RuntimeEnv is base64 string, do we want to expose it same way as base64 to user in apiserver or just ask user to type json string and we can encode it?

File: apiserver/pkg/util/job.go [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+	}
+
+	rayJob := &rayalphaapi.RayJob{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        apiJob.Name,
+			Namespace:   apiJob.Namespace,
+			Labels:      buildRayJobLabels(apiJob),
+			Annotations: buildRayJobAnnotations(apiJob),
+		},
+		Spec: rayalphaapi.RayJobSpec{
```
Comment (Jeffwan) [inline apiserver/pkg/util/job.go:30]: We need to expose TTL setting as well. Check latest job definition

File: apiserver/pkg/util/job.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+			Entrypoint: apiJob.Entrypoint,
+			Metadata:  apiJob.Metadata,
+			RuntimeEnv: apiJob.RuntimeEnv,
+			ShutdownAfterJobFinishes: apiJob.ShutdownAfterJobFinishes,
+			JobId: apiJob.JobId,
+			RayClusterSpec: clusterSpec,
+			ClusterSelector: apiJob.ClusterSelector,
+		},
+		Status: rayalphaapi.RayJobStatus{},
```
Comment (Jeffwan) [inline apiserver/pkg/util/job.go:39]: we can delete 39 for new job. It will be created automatically

File: proto/job.proto [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+  string runtime_env = 6;
+  // If jobId is not set, a new jobId will be auto-generated.
+  string job_id = 7;
+  // If set to true, the rayCluster will be deleted after the rayJob finishes
+  bool ShutdownAfterJobFinishes = 8;
+  // The label selector to choose exiting clusters
+  map<string, string> cluster_selector = 9;
+  // The cluster template
+  ClusterSpec cluster_spec = 10;
+  // logs
```
Comment (Jeffwan) [inline proto/job.proto:125]: we need to add some job related metadata like start_time and end_time here.

File: apiserver/pkg/util/job.go [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+	labels[RayClusterUserLabelKey] = job.User
+	labels[KubernetesApplicationNameLabelKey] = ApplicationName
+	labels[KubernetesManagedByLabelKey] = ComponentName
+	return labels
+}
+
+func buildRayJobAnnotations(cluster *api.RayJob) map[string]string {
+	annotations := map[string]string{}
+	return annotations
+}
```
Comment (Jeffwan) [inline apiserver/pkg/util/job.go:63]: minor: empty line needed

File: apiserver/pkg/manager/resource_manager.go [hunk 13]
```diff
@@ -63,7 +71,7 @@ func (r *ResourceManager) getKubernetesNamespaceClient() clientv1.NamespaceInter
 // clusters
 func (r *ResourceManager) CreateCluster(ctx context.Context, apiCluster *api.Cluster) (*v1alpha1.RayCluster, error) {
 	// populate cluster map
-	computeTemplateDict, err := r.populateComputeTemplate(ctx, apiCluster)
```
Comment (Jeffwan) [inline apiserver/pkg/manager/resource_manager.go:66]: just like to confirm, these refactor are just for reuse, right?

File: apiserver/pkg/manager/resource_manager.go [hunk 14]
```diff
@@ -181,6 +189,101 @@ func (r *ResourceManager) DeleteCluster(ctx context.Context, clusterName string,
...
+	if apiJob.ClusterSpec != nil && len(apiJob.ClusterSelector) == 0 {
+		// populate cluster map
+		computeTemplateMap, err = r.populateComputeTemplate(ctx, apiJob.ClusterSpec, jobNamespace)
+		if err != nil {
+			return nil, util.NewInternalServerError(err, "Failed to populate compute template for (%s/%s)", jobNamespace, apiJob.JobId)
+		}
+	}
+
+	// convert *api.Cluster to v1alpha1.RayCluster
+	rayJob := util.NewRayJob(apiJob, computeTemplateMap)
```
Comment (Jeffwan) [inline apiserver/pkg/manager/resource_manager.go:205]: can not we build directly from `apiJob`? any reason to populate `computeTemplateMap` separately?

File: apiserver/pkg/server/job_server.go [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+// TODO(basasuya) add event outputs
+type RayJobServer struct {
+	resourceManager *manager.ResourceManager
+	api.UnimplementedRayJobServiceServer
+}
+
+
+func (s *RayJobServer) CreateRayJob(ctx context.Context, request *api.CreateRayJobRequest) (*api.RayJob, error) {
+	// use the namespace in the request to override the namespace in the job definition
+	request.Job.Namespace = request.Namespace
```
Comment (Jeffwan) [inline apiserver/pkg/server/job_server.go:31]: for creation, let's do a validation first. check how raycluster did and we can raise some 400 issues.

File: apiserver/cmd/main.go (hunk 1 above)
Comment (Basasuya) [reply apiserver/cmd/main.go]: fixed

File: apiserver/pkg/client/job.go (hunk 2 above)
Comment (Basasuya) [reply apiserver/pkg/client/job.go]: fixed

File: apiserver/pkg/manager/client_manager.go (hunk 3 above)
Comment (Basasuya) [reply apiserver/pkg/manager/client_manager.go]: fixed

File: apiserver/pkg/model/converter.go (hunk 5 above)
Comment (Basasuya) [reply apiserver/pkg/model/converter.go:129]: fixed, add name, namespace in job proto

File: apiserver/pkg/manager/resource_manager.go (hunk 4 above)
Comment (Basasuya) [reply apiserver/pkg/manager/resource_manager.go:35]: added.
BTW, should we use ResourceManagerInterface in each component?

File: apiserver/pkg/util/job.go (hunk 9 above)
Comment (Basasuya) [reply apiserver/pkg/util/job.go:30]: added ttl in job proto

File: apiserver/pkg/util/job.go (hunk 10 above)
Comment (Basasuya) [reply apiserver/pkg/util/job.go]: fixed

File: proto/job.proto (hunk 6 above)
Comment (Basasuya) [reply proto/job.proto]: I would remove logs in job proto temporarily

File: apiserver/pkg/manager/resource_manager.go (hunk 13 above)
Comment (Basasuya) [reply apiserver/pkg/manager/resource_manager.go:66]: yes

File: apiserver/pkg/manager/resource_manager.go (hunk 14 above)
Comment (Basasuya) [reply apiserver/pkg/manager/resource_manager.go:206]: computeTemplateMap is used to construct clusterSpec

File: proto/job.proto (hunk 11 above)
Comment (Basasuya) [reply proto/job.proto]: added

File: proto/job.proto [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+  // If set to true, the rayCluster will be deleted after the rayJob finishes
+  bool ShutdownAfterJobFinishes = 8;
+  // The label selector to choose exiting clusters
+  map<string, string> cluster_selector = 9;
+  // The cluster template
+  ClusterSpec cluster_spec = 10;
+  // TTLSecondsAfterFinished is the TTL to clean up RayCluster.
+  int32 ttl_seconds_after_finished = 11;
+  // Output. The time that the job created.
+  google.protobuf.Timestamp created_at = 12;
```
Comment (Jeffwan) [inline proto/job.proto:129]: Need job status and message as well?

File: apiserver/pkg/manager/resource_manager.go (hunk 14 above)
Comment (Jeffwan) [reply apiserver/pkg/manager/resource_manager.go:206]: this is minor and ok.

File: apiserver/pkg/server/job_server.go (hunk 15 above)
Comment (Jeffwan) [reply apiserver/pkg/server/job_server.go:33]: @Basasuya This is necessary to reduce error creation in cluster. Let's add some validation here.

File: proto/job.proto [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+  // Required input field. Unique job name provided by user.
+  string name = 1;
+  // Required input field. job namespace provided by user
+  string namespace = 2;
+  // Required field. This field indicates the user who owns the job.
+  string user = 3;
+  // The entrypoint of the RayJob
+  string entrypoint = 4;
+  // Metadata is data to store along with this job.
+  map<string, string> metadata = 5;
```
Comment (Jeffwan) [inline proto/job.proto:115]: `metadata` currently is Ray internal job's metadata information. In our downstream, we may need to add some other information like datacenter, queue etc. But this field currently is only used to annodate ray job. We also want to annotate RayCluster.  Let's discuss offline on this.

Comment (Basasuya) [reply proto/job.proto:115]: ok, i would just paste it to the annotation in the ObjectMeta.

File: proto/job.proto (hunk 16 above)
Comment (Basasuya) [reply proto/job.proto:129]: fixed. Add jobStatus, JobDeploymentStatus, message in new proto

**Comment by Jeffwan** [comment]
Let's run this command to fix the lint issue

```
goimports -w apiserver/pkg/model/converter.go
```

File: proto/job.proto (hunk 17 above)
Comment (Jeffwan) [reply proto/job.proto:115]: that sounds good to me

File: apiserver/pkg/util/job.go (hunk 8 above)
Comment (Basasuya) [reply apiserver/pkg/util/job.go]: added, now user could write the runtime env like this:
```yaml
"runtimeEnv": "{'pip': ['requests==2.26.0','pendulum==2.1.2'],'env_vars': {'counter_name': 'test_counter'}}'",
```

File: apiserver/pkg/server/job_server.go (hunk 15 above)
Comment (Basasuya) [reply apiserver/pkg/server/job_server.go:37]: after manual test, job create/list/delete would be OK. but after ray cluster create, seems job can not run through.

**Comment by Basasuya** [comment]
> Let's run this command to fix the lint issue
> 
> ```
> goimports -w apiserver/pkg/model/converter.go
> ```
@Jeffwan fixed

File: apiserver/pkg/server/job_server.go (hunk 15 above)
Comment (Jeffwan) [reply apiserver/pkg/server/job_server.go:37]: I feel that's a job configuration issue. It works on my side and let's discuss this issue offline.

**Comment by Jeffwan** [comment]
Overall looks good to me. I will leave this change for a while. after v0.3.0 official release, we can merge it to avoid conflicts (fix on master can not be easily cherry-picked into v0.3.0)

**Review by Jeffwan — APPROVED** [review]

**Pull request merged**